src/main/java/dataStructures/hashSet/openAddressing/HashSet.java

```diff
@@ -99,19 +103,28 @@ public boolean add(T element) {
      */
     public boolean remove(T element) {
         // If load factor falls below 0.25 and still above minimum size (16), shrink the hashset by half.
-        if (this.size() <= this.capacity() * 0.25 && this.capacity() / 2 >= INITIAL_CAPACITY) {
+        if (size() <= capacity() * 0.25 && capacity() / 2 >= INITIAL_CAPACITY) {
             resize(this.capacity() / 2);
         }
 
-        int bucketIndex = this.search(element);
-        // If the index returned by the probe function contains an empty bucket, then the element is not present in
-        // the set.
-        if (bucketIndex == ELEMENT_NOT_FOUND) {
-            return false;
+        int collisions = 0;
+        while (collisions < capacity()) {
+            int bucketIndex = hashFunction(element, collisions);
+
+            // Element is not removed, because it is not in the Set.
+            if (isNullBucket(bucketIndex)) {
+                return false;
+            }
+
+            if (buckets[bucketIndex].equals(element)) {
+                buckets[bucketIndex] = tombstone(); // Mark the current bucket with a Tombstone.
+                size--;
+                return true;
+            }
+
+            collisions++;
```

Comment:
It may seem that this loop is repeated across `add`, `remove` and `search`. However, each of them is implemented differently in the way they handle buckets containing deleted/null elements.

For `contains`, the loop returns false immediately upon encountering a null element, which indicates the absence of the element. Otherwise, it checks if the element contained (Tombstone or an actual element) is equal to the target element.

For `remove`, the difference is that if the element is found, it is replaced with Tombstone.

For `add`, it first calls `contains`, returning false if the element is already in the set. Else, it finds the first empty bucket - defined as a bucket that contains either `null`, or a `Tombstone`,  then assigns it to the element.

Reply:
A possible, different implementation is that the loop in `contains` and `remove` is abstracted out, and returns the bucketIndex instead, like the earlier implementation, but this introduces a need to differentiate between a valid bucketIndex (contains an element) against an invalid bucketIndex (contains null/tombstone). As such, I chose not to do so, since this introduces unnecessary complexity.